src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentMarkupEncodingPass.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Razor.Language.Components
+{
+    internal class ComponentMarkupEncodingPass : ComponentIntermediateNodePassBase, IRazorOptimizationPass
+    {
+        // Runs after ComponentMarkupBlockPass
+        public override int Order => 10010;
+
+        protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
+        {
+            if (!IsComponentDocument(documentNode))
+            {
+                return;
+            }
+
+            if (documentNode.Options.DesignTime)
+            {
+                // Nothing to do during design time.
+                return;
+            }
+
+            var rewriter = new Rewriter();
+            rewriter.Visit(documentNode);
+        }
+
+        private class Rewriter : IntermediateNodeWalker
+        {
+            // Markup content in components are rendered in one of the following two ways,
+            // AddContent - encodes the content and renders it as plain text (low perf impact)
+            // AddMarkupContent - renders the content directly as markup (high perf impact)
```

Comment:
I'd correct this a bit. 

AddContent doesn't encode the content - it's inserted directly into the DOM in a way that is safe. When used with prerendering we must encode it.